README.md
# Dopefed VCV Rack Plugin

More simple But useful modules for VCV Rack by Strum
2018-19

strum@softhome.net

https://github.com/Strum

Thanks to Andrew Belt and everyone else who has contributed to VCV Rack

## About

This plugin features 3 simple utility modules

SwitchedAdder is a Dual summing mixer with on/off/invert switches 

SwitchedMux is a 8 to 1 mux with manual selector switch

SwitchMatrix is a 10 * 10 3 mode switching matrix

## Usage
https://github.com/Strum/dopefed/wiki/Dopefed-Plugin---User-manual

## Contributing
Post a message on the repo with any suggestions, improvements or bugs or if you have code and want to contribute it

## License
[BSD3]((https://opensource.org/licenses/BSD-3-Clause)

CHANGELOG.md
Dopefed VCV Rack plugin change log

Version 1.0.0
	Added CHANGELOG.md
	Updated URLs in plugin.json
	Edited README.md
	Added mode switches to json rouines in switchmatrix

src/dopefed.hpp
///////////////////////////////////////////////////
//
//   Dopefed Plugin for VCV Version 1
//
//   Strum 2018-19
//   strum@softhome.net
//
///////////////////////////////////////////////////

#include "rack.hpp"

using namespace rack;

// Forward-declare the Plugin, defined in Template.cpp
extern Plugin *pluginInstance;

// Forward-declare each Model, defined in each module source file
extern Model *modelSwitchedAdder;
extern Model *modelSwitchedMux;
extern Model *modelSwitchMatrix;
extern Model *modelQuantiser2;

//// Gui Components

// Ports

struct InPort : SvgPort
{
	InPort()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/InPort.svg")));
	}
};

struct OutPort : SvgPort
{
	OutPort()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/OutPort.svg")));
	}
};

struct GateInPort : SvgPort
{
	GateInPort()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/GateInPort.svg")));
	}
};

struct GateOutPort : SvgPort
{
	GateOutPort()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/GateOutPort.svg")));
	}
};

struct CVInPort : SvgPort
{
	CVInPort()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/CVInPort.svg")));
	}
};

struct CVOutPort : SvgPort
{
	CVOutPort()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/CVOutPort.svg")));
	}
};

struct ThreeWaySwitch : SvgSwitch
{
	ThreeWaySwitch()
	{
		addFrame(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/Three_0.svg")));
		addFrame(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/Three_1.svg")));
		addFrame(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/Three_2.svg")));
	}
};

struct VThreeWaySwitch : SvgSwitch
{
	VThreeWaySwitch()
	{
		addFrame(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/VThree_0.svg")));
		addFrame(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/VThree_1.svg")));
		addFrame(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/VThree_2.svg")));
	}
};

// Knobs

struct LrgKnob : RoundKnob
{
	LrgKnob()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/LrgKnob.svg")));
		box.size = Vec(45,45);
	}
};

struct MedKnob : RoundKnob
{
	MedKnob()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/MedKnob.svg")));
		box.size = Vec(32,32);
	}
};

struct SmlKnob : RoundKnob
{
	SmlKnob()
	{
		setSvg(APP->window->loadSvg(asset::plugin(pluginInstance, "res/components/SmlKnob.svg")));
		box.size = Vec(25,25);
	}
};

/// lights

struct dopefedLight : GrayModuleLightWidget
{
	dopefedLight() { bgColor = nvgRGB(0x40, 0x40, 0x40); }
};
struct RedLED : dopefedLight
{
 	RedLED() { addBaseColor(nvgRGB(0xff, 0x00, 0x00)); }
};

struct BlueLED : dopefedLight
{
 	BlueLED() { addBaseColor(nvgRGB(0x00, 0x47, 0x7e)); }
};

struct OrangeLED : dopefedLight
{
 	OrangeLED() { addBaseColor(nvgRGB(0xff, 0xA5, 0x00)); }
};

template <typename BASE>
 struct TinyLight : BASE
 {
 	TinyLight()	{ this->box.size = Vec(4, 4);}
 };

template <typename BASE>
 struct SmlLight : BASE
 {
 	SmlLight() { this->box.size = Vec(8, 8); }
 };

template <typename BASE>
 struct MedLight : BASE
 {
 	MedLight() { this->box.size = Vec(10, 10); }
 };

src/Quantiser2.cpp
///////////////////////////////////////////////////
//
//   Dopefed Plugin
//   Quantiser2
//
//   Strum 2017-19
//   strum@softhome.net
//
//////////////////////////////////////////////////////////////////////////////////////////////////////

#include "dopefed.hpp"

/////////////////////////////////////////////////
struct Quantiser2 : Module
{
	enum ParamIds
  {
      PITCH_PARAM,
      BUTTON_PARAM,      
      NUM_PARAMS = BUTTON_PARAM + 12
	};
	enum InputIds
  {
      INPUT,
      PITCH_INPUT,
      NUM_INPUTS
	};
	enum OutputIds
  {
      OUTPUT,
      REF_OUT,
      NUM_OUTPUTS = REF_OUT + 12
	};
  enum LightIds
  {
		BUTTON_LIGHTS,
    OUTPUT_LIGHTS = BUTTON_LIGHTS + 12,
		NUM_LIGHTS = OUTPUT_LIGHTS + 12
	};

  dsp::SchmittTrigger button_triggers[12];
  
  bool button_states[12] = {true,true,true,true,true,true,true,true,true,true,true,true};
  float quantised[16] = {0.0};
  //bool found = false;
  //int last_found = 0;
   
  Quantiser2()
  {
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);

    configParam(Quantiser2::PITCH_PARAM, -6.5, 6.5, 0.0, "");
    for (int i = 0; i < 12; ++i)
    {
      configParam(Quantiser2::BUTTON_PARAM + i, 0.0, 1.0, 0.0, ""); 
    }    
  }

	void process(const ProcessArgs& args) override;
  
  json_t *dataToJson() override
  {
		json_t *rootJ = json_object();
    
    // button states
		json_t *button_statesJ = json_array();
		for (int i = 0; i < 12; i++)
    {
			json_t *button_stateJ = json_integer((int) button_states[i]);
			json_array_append_new(button_statesJ, button_stateJ);
		}
		json_object_set_new(rootJ, "buttons", button_statesJ);    
    return rootJ;
  }
  
  void dataFromJson(json_t *rootJ) override
  {
    // button states
		json_t *button_statesJ = json_object_get(rootJ, "buttons");
		if (button_statesJ)
    {
			for (int i = 0; i < 12; i++)
      {
				json_t *button_stateJ = json_array_get(button_statesJ, i);
				if (button_stateJ)
					button_states[i] = !!json_integer_value(button_stateJ);
			}
		}  
  }
};


/////////////////////////////////////////////////////
void Quantiser2::process(const ProcessArgs& args)
{
  int numChannels = inputs[INPUT].getChannels();

  ////// handle button presses
  for  (int i = 0 ; i < 12 ; i++)
  {
    if (button_triggers[i].process(params[BUTTON_PARAM+i].getValue()))
    {
		  button_states[i] = !button_states[i];
	  }
    lights[BUTTON_LIGHTS + i ].value  = (button_states[i]) ? 1.0 : 0.0;
    lights[OUTPUT_LIGHTS + i].value = 0.0;
  }

  // pitch offset
  float pitch_in = round(inputs[PITCH_INPUT].getVoltage())/12;
  float root_pitch = (pitch_in * (1/12.0)) + (round(params[PITCH_PARAM].getValue()) * (1/12.0)); 
    
  // set reference outputs
  for  (int i = 0 ; i < 12 ; i++)
  {
    outputs[REF_OUT + i].setVoltage(root_pitch + i * (1/12.0));
  }
  
  //////// quantise pitch to chromatic scale
  for (int channel = 0; channel < numChannels; channel++)
  {
    float in = inputs[INPUT].getVoltage(channel);  
    int octave = round(in);
    float octaves_removed = in - 1.0*octave;
    int semitone = round(octaves_removed*12);
    if (semitone < 0)
    { 
      semitone +=12;
      octave -= 1;
    }
    quantised[channel] = root_pitch + 1.0 * octave + semitone/12.0;
      
    // quantise to scale selected by buttons
    if (button_states[semitone])
    {    
      //found = true;    
      outputs[OUTPUT].setVoltage(quantised[channel], channel);
      //lights[OUTPUT_LIGHTS + semitone].value  = 1.0;
    } 
  }
  outputs[OUTPUT].setChannels(numChannels);   
}

//////////////////////////////////////////////////////////////////
struct Quantiser2Widget : ModuleWidget
{
  Quantiser2Widget(Quantiser2 *module)
  {
    setModule(module);

    setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/Quantiser2.svg")));

    int top_row = 40;
    int row_spacing = 25; 
  	
    addParam(createParam<MedKnob>(Vec(62, 15), module, Quantiser2::PITCH_PARAM));
    addInput(createInput<CVInPort>(Vec(63, 45), module, Quantiser2::PITCH_INPUT));
    
    addInput(createInput<CVInPort>(Vec(3, top_row), module, Quantiser2::INPUT));
    addOutput(createOutput<CVOutPort>(Vec(32, top_row), module, Quantiser2::OUTPUT));
    
    for (int i = 0; i < 12 ; i++)
    {  
      addParam(createParam<LEDButton>(Vec(3, top_row + 30 + row_spacing * i), module, Quantiser2::BUTTON_PARAM + i));
  	  addChild(createLight<MedLight<BlueLED>>(Vec(3+5, top_row + 30 + row_spacing * i + 5), module, Quantiser2::BUTTON_LIGHTS + i));
      addChild(createLight<MedLight<BlueLED>>(Vec(30+5, top_row + 30 + row_spacing * i + 5), module, Quantiser2::OUTPUT_LIGHTS + i));
      addOutput(createOutput<CVOutPort>(Vec(63, top_row + 40 + row_spacing * i), module, Quantiser2::REF_OUT + i));    
    }
  }
};

Model *modelQuantiser2 = createModel<Quantiser2, Quantiser2Widget>("Quantiser2");

src/SwitchedMux.cpp
///////////////////////////////////////////////////
//
//   Dopefed Plugin
//   8 to 1 Mux with Manual selector switch
//
//   Strum 2018-19
//   strum@softhome.net
//
///////////////////////////////////////////////////

#include "dopefed.hpp"

struct SwitchedMux : Module
{
	enum ParamIds
  {
    BUTTON_1,
    NUM_PARAMS = BUTTON_1 + 8
	};  
	enum InputIds
  {    
    SIG_INPUT,		  
		NUM_INPUTS = SIG_INPUT + 8
	};
	enum OutputIds
  {
    OUTPUT,    
		NUM_OUTPUTS
	};
  enum LightIds
  {
    BUTTON_LEDS,
    NUM_LIGHTS = BUTTON_LEDS + 8
  };  
    
  dsp::SchmittTrigger button_triggers[8];
  int selected_input = 0;
  bool is_new = true;
  
  SwitchedMux()
  {
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
    for (int i = 0; i < 8; ++i)
    {
      configParam(SwitchedMux::BUTTON_1 + i, 0.0f, 1.0f, 0.0f, "");
    }
  };
  
  void process(const ProcessArgs& args) override;   
  
  json_t *dataToJson() override
  {
    json_t *rootJ = json_object();
    
    json_object_set_new(rootJ, "selected", json_integer(selected_input));
     
    return rootJ;
  };
  
  void dataFromJson(json_t *rootJ) override
  {
   json_t *selected_inputJ = json_object_get(rootJ, "selected");
    if(selected_inputJ) selected_input = json_integer_value(selected_inputJ);
  };  
};

//////////
void SwitchedMux::process(const ProcessArgs& args)
{
  if (is_new)
  {
    is_new = false;
    if (selected_input == 0) lights[BUTTON_LEDS].value = 1.0;
    else lights[BUTTON_LEDS + selected_input].value = 1.0;    
  }

  for ( int i = 0 ; i < 8 ; i ++)
  {
    if (button_triggers[i].process(params[BUTTON_1 + i].getValue()))
    {
      selected_input = i;
      for (int j = 0 ; j < 8 ; j ++)
      {        
        lights[BUTTON_LEDS + j ].value = (j==selected_input)?1.0:0.0;        
      }
    }    
  }
  outputs[OUTPUT].setVoltage(inputs[SIG_INPUT + selected_input].getVoltage());  
}

//////////////////////////////////////////////////////////////
struct SwitchedMuxWidget : ModuleWidget
{
  SwitchedMuxWidget(SwitchedMux *module)
  {
    setModule(module);
    setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/SwitchedMux.svg")));
	
    int spacing = 30; 
    int top_space = 15;    
  
    for (int i = 0; i < 8 ; i++)
    { 
      addParam(createParam<LEDButton>(Vec(7,top_space + spacing * i + 54 ), module, SwitchedMux::BUTTON_1 + i)); 
      addInput(createInput<InPort>(Vec(28, top_space + spacing * i + 50), module, SwitchedMux::SIG_INPUT + i));   	 
      addChild(createLight<MediumLight<GreenLight>>(Vec(11, top_space +  spacing * i + 5 + 53), module, SwitchedMux::BUTTON_LEDS + i));
    }  
    addOutput(createOutput<OutPort>(Vec(18, 333), module, SwitchedMux::OUTPUT)); 
  }  
};

Model *modelSwitchedMux = createModel<SwitchedMux, SwitchedMuxWidget>("SwitchedMux");

src/SwitchedAdder.cpp
///////////////////////////////////////////////////
//
//   Dopefed Plugin
//   switched precision adder
//
//   Strum 2018-19
//   strum@softhome.net
//
///////////////////////////////////////////////////

#include "dopefed.hpp"

struct SwitchedAdder : Module
{
	enum ParamIds
    {
        CH1_IN1_SWITCH,
        CH1_IN2_SWITCH,
        CH1_IN3_SWITCH,
        CH1_IN4_SWITCH,
        CH1_IN5_SWITCH,
        CH2_IN1_SWITCH,
        CH2_IN2_SWITCH,
        CH2_IN3_SWITCH,
        CH2_IN4_SWITCH,
        CH2_IN5_SWITCH,
		NUM_PARAMS
	};  
	enum InputIds
    {
        INPUT_1,
        INPUT_2,
        INPUT_3,
        INPUT_4,
        INPUT_5,
        INPUT_2_1,
        INPUT_2_2,
        INPUT_2_3,
        INPUT_2_4,
        INPUT_2_5,			  
		NUM_INPUTS
	};
	enum OutputIds
    {
		OUTPUT_1,
        OUTPUT_2,
        OUTPUT_1_INV,
        OUTPUT_2_INV,        
		NUM_OUTPUTS
	};
  
    float sum_out_1 = 0.0;
    float sum_out_2 = 0.0;  
  
	SwitchedAdder()
    {
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS);
        configParam(SwitchedAdder::CH1_IN1_SWITCH, 0.0f, 2.0f, 0.0f, "");
        configParam(SwitchedAdder::CH1_IN2_SWITCH, 0.0f, 2.0f, 0.0f, "");
        configParam(SwitchedAdder::CH1_IN3_SWITCH, 0.0f, 2.0f, 0.0f, "");
        configParam(SwitchedAdder::CH1_IN4_SWITCH, 0.0f, 2.0f, 0.0f, "");
        configParam(SwitchedAdder::CH1_IN5_SWITCH, 0.0f, 2.0f, 0.0f, "");
        configParam(SwitchedAdder::CH2_IN1_SWITCH, -1.0f, 1.0f, 0.0f, "");
        configParam(SwitchedAdder::CH2_IN2_SWITCH, -1.0f, 1.0f, 0.0f, "");
        configParam(SwitchedAdder::CH2_IN3_SWITCH, -1.0f, 1.0f, 0.0f, "");
        configParam(SwitchedAdder::CH2_IN4_SWITCH, -1.0f, 1.0f, 0.0f, "");
        configParam(SwitchedAdder::CH2_IN5_SWITCH, -1.0f, 1.0f, 0.0f, "");
    }

	void process(const ProcessArgs& args) override;

    
};

void SwitchedAdder::process(const ProcessArgs& args)
{
  sum_out_1 = 0.0;
  sum_out_2 = 0.0;
  
  for (int i = 0 ; i < 5 ; i++)
  {
    if (params[CH1_IN1_SWITCH + i].getValue() == 0) 
      sum_out_1 -= inputs[INPUT_1 + i].getVoltage();
    else if (params[CH1_IN1_SWITCH + i].getValue() == 2) 
      sum_out_1 += inputs[INPUT_1 + i].getVoltage();
    
    sum_out_2 += inputs[INPUT_2_1 + i].getVoltage() * params[CH2_IN1_SWITCH + i].getValue();
  }
  
  outputs[OUTPUT_1].setVoltage(sum_out_1);
  outputs[OUTPUT_2].setVoltage(sum_out_2);
  outputs[OUTPUT_1_INV].setVoltage(- sum_out_1);
  outputs[OUTPUT_2_INV].setVoltage(- sum_out_2);  
}

////////////////////////////////////////////////////////////////
struct SwitchedAdderWidget : ModuleWidget
{
	SwitchedAdderWidget(SwitchedAdder *module)
    {
        setModule(module);
        setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/SwitchedAdder.svg"))); 
      
        // channel 1
        addInput(createInput<InPort>(Vec(10, 30), module, SwitchedAdder::INPUT_1));
        addInput(createInput<InPort>(Vec(10, 55), module, SwitchedAdder::INPUT_2));
        addInput(createInput<InPort>(Vec(10, 80), module, SwitchedAdder::INPUT_3));
        addInput(createInput<InPort>(Vec(10, 105), module, SwitchedAdder::INPUT_4));
        addInput(createInput<InPort>(Vec(10, 130), module, SwitchedAdder::INPUT_5));

        addParam(createParam<ThreeWaySwitch>(Vec(40, 30), module, SwitchedAdder::CH1_IN1_SWITCH));
        addParam(createParam<ThreeWaySwitch>(Vec(40, 55), module, SwitchedAdder::CH1_IN2_SWITCH));
        addParam(createParam<ThreeWaySwitch>(Vec(40, 80), module, SwitchedAdder::CH1_IN3_SWITCH));
        addParam(createParam<ThreeWaySwitch>(Vec(40, 105), module, SwitchedAdder::CH1_IN4_SWITCH));
        addParam(createParam<ThreeWaySwitch>(Vec(40, 130), module, SwitchedAdder::CH1_IN5_SWITCH)); 

        addOutput(createOutput<OutPort>(Vec(10, 160), module, SwitchedAdder::OUTPUT_1));
        addOutput(createOutput<OutPort>(Vec(33, 160), module, SwitchedAdder::OUTPUT_1_INV)); 
      
        // channel 2
        int offset = 20;
        addInput(createInput<InPort>(Vec(10, 190 + offset), module, SwitchedAdder::INPUT_2_1));
        addInput(createInput<InPort>(Vec(10, 215 + offset), module, SwitchedAdder::INPUT_2_2));
        addInput(createInput<InPort>(Vec(10, 240 + offset), module, SwitchedAdder::INPUT_2_3));
        addInput(createInput<InPort>(Vec(10, 265 + offset), module, SwitchedAdder::INPUT_2_4));
        addInput(createInput<InPort>(Vec(10, 290 + offset), module, SwitchedAdder::INPUT_2_5));	

        addParam(createParam<ThreeWaySwitch>(Vec(40, 190 + offset), module, SwitchedAdder::CH2_IN1_SWITCH));
        addParam(createParam<ThreeWaySwitch>(Vec(40, 215 + offset), module, SwitchedAdder::CH2_IN2_SWITCH));
        addParam(createParam<ThreeWaySwitch>(Vec(40, 240 + offset), module, SwitchedAdder::CH2_IN3_SWITCH));
        addParam(createParam<ThreeWaySwitch>(Vec(40, 265 + offset), module, SwitchedAdder::CH2_IN4_SWITCH));
        addParam(createParam<ThreeWaySwitch>(Vec(40, 290 + offset), module, SwitchedAdder::CH2_IN5_SWITCH));

        addOutput(createOutput<OutPort>(Vec(10, 315 + offset), module, SwitchedAdder::OUTPUT_2));
        addOutput(createOutput<OutPort>(Vec(33, 315 + offset), module, SwitchedAdder::OUTPUT_2_INV));
    }
};

Model *modelSwitchedAdder = createModel<SwitchedAdder, SwitchedAdderWidget>("SwitchedAdder");

src/SwitchMatrix.cpp
///////////////////////////////////////////////////////////////////
//
//  Dopefed Plugin
//  Switch Matrix 
//
//  Strum 2018-19
//  strum@softhome.net
//
///////////////////////////////////////////////////////////////////

#include "dopefed.hpp"


struct SwitchMatrix : Module
{
	enum ParamIds
  {
    MODE_SEL,
    SWITCHES = MODE_SEL + 10,
		NUM_PARAMS = SWITCHES + 100
	};  
	enum InputIds
  {		
		INPUTS,
    NUM_INPUTS = INPUTS + 10
	};
	enum OutputIds
  {
		OUTPUTS,    
		NUM_OUTPUTS = OUTPUTS + 10
	};
  enum LightIds
  {
		SWITCH_LIGHTS,
    NUM_LIGHTS = SWITCH_LIGHTS + 100
	};  

  dsp::SchmittTrigger switch_triggers[10][10];

  bool switch_states[10][10] = 
  {{0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0},
  {0,0,0,0,0,0,0,0,0,0}};   
  
  float input_values[10] = {0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0};
  float sums[10] = {0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0}; 
  int selected_in[10] = {0,0,0,0,0,0,0,0,0,0};
  int mode[10] = {0,0,0,0,0,0,0,0,0,0};

	SwitchMatrix()
  {
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
    
    for (int i = 0; i < 10 ; i++)
    {  
      configParam(SwitchMatrix::MODE_SEL + i, 0.0f, 2.0f, 0.0f, "");    
      for (int j = 0; j < 10 ; j++)
      {        
        configParam(SwitchMatrix::SWITCHES + i * 10 + j, 0.0f, 1.0f, 0.0f, "");
      }
    }
  }

	void process(const ProcessArgs& args) override;
  
  json_t *dataToJson() override
  {
		json_t *rootJ = json_object();
    
    // button states
		json_t *button_statesJ = json_array();
    json_t *mode_statesJ = json_array();
    for (int i = 0; i < 10; i++)
    {
      json_t *mode_stateJ = json_integer(mode[i]);
      json_array_append_new(mode_statesJ, mode_stateJ);
    }
    json_object_set_new(rootJ, "modes", mode_statesJ);
		for (int i = 0; i < 10; i++)
    {
			for (int j = 0; j < 10; j++)
      {
        json_t *button_stateJ = json_integer((int) switch_states[i][j]);
			  json_array_append_new(button_statesJ, button_stateJ);
      }
		}
		json_object_set_new(rootJ, "buttons", button_statesJ);
    
    return rootJ;
  }
  
  void dataFromJson(json_t *rootJ) override
  {
    // button states
    json_t *mode_statesJ = json_object_get(rootJ, "modes");
		json_t *button_statesJ = json_object_get(rootJ, "buttons");

    for (int i = 0; i < 10; i++)
    {
      json_t *mode_stateJ = json_array_get(mode_statesJ, i);
      if (mode_stateJ)
        mode[i] = json_integer_value(mode_stateJ);
    }
		if (button_statesJ)
    {
			for (int i = 0; i < 10; i++)
      {
        for (int j = 0; j < 10; j++)
        {
				  json_t *button_stateJ = json_array_get(button_statesJ, i*10 + j);
				  if (button_stateJ)
					  switch_states[i][j] = !!json_integer_value(button_stateJ);
        }
      }
		}  
  }  
};

//////////////////////////////////////////////////////////////
////// Step function
void SwitchMatrix::process(const ProcessArgs& args)
{
	// zero sums
  for ( int i = 0 ; i < 10 ; i++)
  {
    sums[i] = 0.0;
    if (mode[i] != params[MODE_SEL + i].getValue() )
    {
      mode[i] = params[MODE_SEL + i].getValue();
      if (mode[i] == 2) 
      {
        selected_in[i] = 0;
        for (int j = 1 ; j < 10 ; j++)
        {          
          switch_states[i][j] = false ;
          lights[SWITCH_LIGHTS + i * 10 + j ].value = 0.0f;
        }
        switch_states[i][0] = true ;
        lights[SWITCH_LIGHTS + i * 10 ].value = 1.0f;
      } else
      {
        for (int j = 0 ; j < 10 ; j++)
        {          
          switch_states[i][j] = false ;
          lights[SWITCH_LIGHTS + i * 10 + j ].value = 0.0f;
        }
      }
    }
  }
  
  // deal with buttons
    for (int col = 0 ; col < 10 ; col++)
  {
    for (int row = 0 ; row < 10 ; row++)
    {
      if (params[MODE_SEL + col].getValue() != 2)
      {
          if (switch_triggers[col][row].process(params[SWITCHES+col*10 + row].getValue()))
          {
            switch_states[col][row] = !switch_states[col][row];
          }
          lights[SWITCH_LIGHTS + col * 10 + row].value  = (switch_states[col][row]) ? 1.0 : 0.0;
      }
      else
      {
        if (switch_triggers[col][row].process(params[SWITCHES+col*10 + row].getValue()))
        {
          for (int i = 0 ; i  < 10 ; i++)
          {
           switch_states[col][i] = false ;
           lights[SWITCH_LIGHTS + col * 10 + i ].value = 0.0f;
          }
          switch_states[col][row] = true;
          selected_in[col] = row;
          lights[SWITCH_LIGHTS + col * 10 + row ].value = 1.0f;
        }           
      }
    }       
  }

  // get inputs
  for (int i = 0 ; i < 10 ; i++)
  {
    input_values[i] = inputs[INPUTS + i].getVoltage();
  }
  
  // process  inputs  
  for (int i = 0 ; i < 10 ; i++)
  {   
    // sum mode
    if (params[MODE_SEL + i].getValue() == 0)
    {
      for (int j = 0 ; j < 10 ; j++)
      {
        if (switch_states[i][j]) sums[i] += input_values[j];
      }
    }
    // or mode
    if (params[MODE_SEL + i].getValue() == 1)
    {
      for (int j = 0 ; j < 10 ; j++)
      {
        if (switch_states[i][j] && (inputs[INPUTS + j].getVoltage() > 0.0f)) sums[i] = 10.0f;
      }
    }
    // mux mode
    if (params[MODE_SEL + i].getValue() == 2)
      sums[i] = input_values[selected_in[i]];

    // set output
    outputs[OUTPUTS + i].setVoltage(sums[i]);    
  }  
} // end step loop

///// Gui
struct SwitchMatrixWidget : ModuleWidget
{
  SwitchMatrixWidget(SwitchMatrix *module)
  {
    setModule(module);
  
    setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/SwitchMatrix.svg")));

    int top_row = 75;
    int row_spacing = 25; 
    int column_spacing = 25;

  	for (int i = 0 ; i < 10 ; i++)
    {
      addParam(createParam<VThreeWaySwitch>(Vec(35 + i * column_spacing , 30), module, SwitchMatrix::MODE_SEL + i));
      addInput(createInput<InPort>(Vec(3, i * row_spacing + top_row), module, SwitchMatrix::INPUTS + i));  
      addOutput(createOutput<OutPort>(Vec(33 + i * column_spacing , top_row + 10 * row_spacing), module, SwitchMatrix::OUTPUTS + i));
      for(int j = 0 ; j < 10 ; j++ )
      {
        addParam(createParam<LEDButton>(Vec(35 + column_spacing * i, top_row + row_spacing * j), module, SwitchMatrix::SWITCHES + i * 10 + j ));
        addChild(createLight<MedLight<BlueLED>>(Vec(35 + column_spacing * i + 4, top_row + row_spacing * j + 4), module, SwitchMatrix::SWITCH_LIGHTS  + i * 10 + j ));
      }
  	}
  }  
};

Model *modelSwitchMatrix = createModel<SwitchMatrix, SwitchMatrixWidget>("SwitchMatrix");